### PR TITLE
[Storage] Support Immunablity Policy in Official powershell release

### PR DESCRIPTION
**Pull request by blueww**
## Description

To support WORM (Immunablity Policy) in powershell official release, the PR for storage branch is: https://github.com/Azure/azure-powershell/pull/6151
The design pass review on
https://github.com/Azure/azure-powershell-pr/wiki/%5BAzureRM.Storage%5D-ImmutabilityPolicy

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Storage/Commands.Management.Storage.Test/ScenarioTests/StorageBlobTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        public StorageBlobTests(ITestOutputHelper output)
+        {
+            XunitTracingInterceptor.AddToContext(new XunitTracingInterceptor(output));
+        }
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TestStorageBlobContainer()
+        {
+            TestController.NewInstance.RunPsTest("Test-StorageBlobContainer");
```
Comment (maddieclayton) [inline src/ResourceManager/Storage/Commands.Management.Storage.Test/ScenarioTests/StorageBlobTests.cs]: If you pull in latest from preview you will see that a new parameter has been added - this should be
```
TestController.NewInstance.RunPsTest("Test-StorageBlobContainer", interceptor);
```

And you will need to create a local XunitTracingInterceptor and add that at line 28 (see other test cs files for example).

File: src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj [hunk 2]
```diff
@@ -59,18 +60,22 @@
...
+    </None>
+    <None Include="..\..\Common\Commands.ScenarioTests.ResourceManager.Common\AzureRM.Resources.ps1">
+      <Link>ScenarioTests\AzureRM.Resources.ps1</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
     <None Include="ScenarioTests\StorageAccountTests.ps1">
-      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Include="SessionRecords\Microsoft.Azure.Commands.Management.Storage.Test.ScenarioTests.StorageAccountTests\TestGetAzureStorageAccount.json">
```
Comment (maddieclayton) [inline src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: You need to add this back in to get the test to pass

**Comment by maddieclayton** [comment]
@blueww Please also fix the static analysis errors: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/7020/artifact/src/Package/SignatureIssues.csv/*view*/

File: src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj (hunk 2 above)
Comment (blueww) [reply src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: change the file by adding all *.json in the session record folder.

**Comment by blueww** [comment]
@maddieclayton 
I have fixed the Unittest and the StaticAnalysis error.
BTW, why we force the cmdlets must have output, and need 'PassThru' for the bool output.

**Comment by blueww** [comment]
@maddieclayton 

The .net build failed in the last. Seems not related with my code change, would you please help to look?

> 23:00:46 <<<<
> 23:00:46 
> 23:00:46 ERROR
> 23:00:46 
> 23:00:46 Credential scanner match found.
> 23:00:46 Details can be found in the CredScanIssues.csv file.
> 23:00:46 Path to the file: artifact/src/Package/CredentialScannerOutput/CredScanIssues.csv
> 23:00:46 For false positives, follow the instructions on this doc: https://github.com/Azure/adx-documentation-pr/blob/master/engineering/credscan-ps.md
> 23:00:46 For more information, please read the 'How should I respond to reports of incorrectly handled secrets?' section 
> 23:00:46 of this doc: https://microsoft.sharepoint.com/teams/AzureSecurityCompliance/Security/SitePages/SecretsManagement.aspx.
> 23:00:46 
> 23:00:46 >>>>
> 23:00:46 ==== CredScanner: Exit script
> 

The netcore build failed since following error, but I don't find network project reference Microsoft.Azure.Management.Storage, how whould I fix it？(Actually, I have upgrade all netcore project reference Microsoft.Azure.Management.Storage 7.1 to reference  8.1)


[xUnit.net 00:00:16.13]     Commands.Network.Test.ScenarioTests.ApplicationGatewayTests.TestApplicationGatewayCRUD [FAIL]

Failed   Commands.Network.Test.ScenarioTests.ApplicationGatewayTests.TestApplicationGatewayCRUD

Error Message:

 System.Management.Automation.RuntimeException : Test failed due to a non-empty error stream. First error: PowerShell Error Record: Could not load file or assembly 'Microsoft.Azure.Management.Storage, Version=8.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

**Comment by MiYanni** [comment]
@azuresdkci Retest this please

**Comment by blueww** [comment]
@MiYanni 
The netcore build still fail, I thinks it might related with SRP SDK is referenced in azure-powershell-common:
https://github.com/Azure/azure-powershell-common/blob/ef926a1e40814b79c7d14f0a8afc235c93523428/src/Storage.Management/Storage.Management.Netcore.csproj#L39 
https://github.com/Azure/azure-powershell-common/blob/ef926a1e40814b79c7d14f0a8afc235c93523428/src/Storage/Storage.Netcore.csproj#L41

I will try to upgrade SRPSDK in azure-powershell-common (still use the number 1.0.96), the XSCL will still not 9.0 (not upgrade to 9.3) as this PR doesn't include XSCL upgrade code.

**Comment by MiYanni** [comment]
@blueww I merged the OAuth PR. So, this PR needs to be updated (conflicts resolved) so that it can build properly.

**Comment by blueww** [comment]
@MiYanni 

I have fixed the merge conflict, and fixed the cmdlet name for new cmdlets.
But the netcore build still fail as following error, any idea?


>  System.Management.Automation.CommandNotFoundException : The term 'New-AzureRmStorageContainer' is not recognized as the name of a cmdlet, function, script file, or operable program.

**Comment by MiYanni** [comment]
@blueww There is an issue with the prefixing for Storage cmdlet names right now. @maddieclayton will look into making the fix for your branch here.

**Comment by blueww** [comment]
Thanks for the helping!
When can it been fixed?
Let me know if need any help from me.

Br,
Wei

在 2018年9月7日，上午1:53，Michael Yanni <notifications@github.com<mailto:notifications@github.com>> 写道：


@blueww<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fblueww&data=02%7C01%7Cweiwei%40microsoft.com%7C7a0ec88b78d644c4048e08d61421b69b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636718532358149281&sdata=%2FVv8BVV5jwysEZFKASXCJPK3bErinRWRCIWwCvhSm0U%3D&reserved=0> There is an issue with the prefixing for Storage cmdlet names right now. @maddieclayton<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmaddieclayton&data=02%7C01%7Cweiwei%40microsoft.com%7C7a0ec88b78d644c4048e08d61421b69b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636718532358149281&sdata=IKSbeW0jYuMLDz92ffi9qBB6uzupt7kpJ%2Bk3Xj2GD30%3D&reserved=0> will look into making the fix for your branch here.

**Comment by blueww** [comment]
On Demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/1083/
ps-sign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/1187/
psnetsign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/psnetsign/161/

**Comment by MiYanni** [comment]
@blueww Ah. psnetsign ran on worker 78s. It might not have the latest PowerShellGet. Maddie fixed that issue with this PR, https://github.com/Azure/azure-powershell/pull/7154, but I think that worker isn't updated properly. But, everything else looks good. I'll review this PR in a few hours from now.

**Comment by MiYanni** [comment]
psnetsign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/psnetsign/176/

**Comment by MiYanni** [comment]
demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/1087/
ps-sign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/1216/
psnetsign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/psnetsign/178/

**Comment by blueww** [comment]
@MiYanni , @maddieclayton 
Thanks very much for all the work to make the PR merged!

**Pull request merged**